dnsnameresolver.go

```diff
@@ -0,0 +1,291 @@
+			dnsName := string(resolverObj.Spec.Name)
+			// Check if the DNS name is wildcard or regular.
+			if isWildcard(dnsName) {
+				// If the DNS name is wildcard, add the details of the DNSNameResolver
+				// object to the wildcardDNSInfo map.
+				resolver.wildcardMapLock.Lock()
+				dnsInfoMap, exists := resolver.wildcardDNSInfo[dnsName]
+				// If details of DNS name and the DNSNameResolver objects already exist
+				// then check if the existing information match with the current one.
+				// Otherwise, don't proceed.
+				if exists && dnsInfoMap[resolverObj.Namespace] != resolverObj.Name {
+					resolver.wildcardMapLock.Unlock()
+					return
+				}
+				if !exists {
+					dnsInfoMap = make(namespaceDNSInfo)
+				}
+				dnsInfoMap[resolverObj.Namespace] = resolverObj.Name
+				resolver.wildcardDNSInfo[dnsName] = dnsInfoMap
+				resolver.wildcardMapLock.Unlock()
+			} else {
+				// If the DNS name is regular, add the details of the DNSNameResolver
+				// object to the regularDNSInfo map.
+				resolver.regularMapLock.Lock()
+				dnsInfoMap, exists := resolver.regularDNSInfo[dnsName]
+				// If details of DNS name and the DNSNameResolver objects already exist
+				// then check if the existing information match with the current one.
+				// Otherwise, don't proceed.
+				if exists && dnsInfoMap[resolverObj.Namespace] != resolverObj.Name {
+					resolver.regularMapLock.Unlock()
+					return
+				}
+				if !exists {
+					dnsInfoMap = make(namespaceDNSInfo)
+				}
+				dnsInfoMap[resolverObj.Namespace] = resolverObj.Name
+				resolver.regularDNSInfo[dnsName] = dnsInfoMap
+				resolver.regularMapLock.Unlock()
+			}
+
+			// Used only in unit tests.
+			if send != nil {
+				send(resolverObj)
+			}
+		},
+		// Delete event.
+		DeleteFunc: func(obj interface{}) {
+			// Get the DNSNameResolver object.
+			resolverObj, ok := obj.(*dnsv1alpha1.DNSNameResolver)
+			if !ok {
+				log.Infof("object not of type DNSNameResolver: %v", obj)
+				return
+			}
+
+			// Check if namespace is configured or not.
+			if !resolver.configuredNamespace(resolverObj.Namespace) {
+				return
+			}
+
+			dnsName := string(resolverObj.Spec.Name)
+			// Check if the DNS name is wildcard or regular.
+			if isWildcard(dnsName) {
+				// If the DNS name is wildcard, delete the details of the DNSNameResolver
+				// object from the wildcardDNSInfo map.
+				resolver.wildcardMapLock.Lock()
+				if dnsInfoMap, exists := resolver.wildcardDNSInfo[dnsName]; exists {
+					// If details of DNS name and the DNSNameResolver objects already exist
+					// then check if the existing information match with the current one.
+					// Otherwise, don't proceed.
+					if dnsInfoMap[resolverObj.Namespace] != resolverObj.Name {
+						resolver.wildcardMapLock.Unlock()
+						return
+					}
+					delete(dnsInfoMap, resolverObj.Namespace)
+					if len(dnsInfoMap) > 0 {
+						resolver.wildcardDNSInfo[dnsName] = dnsInfoMap
+					} else {
+						delete(resolver.wildcardDNSInfo, dnsName)
+					}
```

Comment:
```suggestion
                    if dnsInfoMap[resolverObj.Namespace] == resolverObj.Name {                                
                        delete(resolver.wildcardDNSInfo[dnsName], resolverObj.Namespace)
                        if len(resolver.wildcardDNSInfo[dnsName]) == 0 { 
                            delete(resolver.wildcardDNSInfo, dnsName)
                        }   
                    } 
```
To avoid calling `Unlock` twice and reassigning `dnsInfoMap` back to `regularDNSInfo`.